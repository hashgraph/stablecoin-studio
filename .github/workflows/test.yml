name: test
on:
  pull_request:
    branches: [ main, release/**, feature/** ]
  push:
    branches: [ main, release/** ]
    tags: [ v* ]
jobs:
  test-node:
    name: Unit Tests
    runs-on: ubuntu-latest
    steps:
      - name: Use Node.js TLS 16.x
        uses: actions/setup-node@v1
        with:
          node-version: 16.x
      - name: Checkout repository
        uses: actions/checkout@v2
      - name: Install dependencies
        run: npm i
      - name: Build
        run: npm run build
      - name: Test
        run: npm run test