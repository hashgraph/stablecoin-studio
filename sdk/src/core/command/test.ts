/* eslint-disable @typescript-eslint/no-empty-interface */

interface CommandResponse {}
interface ICommand {}

export class Command<T> implements ICommand {
	private $resultType!: T;
}

export interface ICommandBus {
	execute<X>(command: Command<X>): Promise<X>;
}

export type IInferringCommandHandler<CommandType extends Command<unknown>> =
	CommandType extends Command<infer ResultType>
		? {
				execute(command: CommandType): Promise<ResultType>;
		  }
		: never;

class ConcreteCommandResponse implements CommandResponse {
	constructor(public readonly status: string) {}
}

class ConcreteCommand extends Command<ConcreteCommandResponse> {
	constructor(public readonly value: string) {
		super();
	}
}

class CommandHandler<CommandType extends Command<unknown>>
	implements IInferringCommandHandler<ConcreteCommand>
{
	execute(command: ConcreteCommand): Promise<ConcreteCommandResponse> {
		throw new Error('Method not implemented.');
	}
}

class CommandBus implements ICommandBus {
	execute<X>(command: Command<X>): Promise<X> {
		throw new Error('Method not implemented.');
	}
}

// Test

const bus = new CommandBus();
const res = bus.execute(new ConcreteCommand('1'));
