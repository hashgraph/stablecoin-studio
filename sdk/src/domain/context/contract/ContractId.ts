import { PublicKey as HPublicKey } from '@hashgraph/sdk';
import { ValueObject } from '../../../core/types.js';
import { proto } from '@hashgraph/proto';
import InvalidKeyForContractIdDomainError from './error/InvalidKeyForContractIdDomainError.js';
import Long from 'long';

export default class ContractId extends ValueObject {
	public readonly id: string;

	constructor(id: string) {
		super();
		this.id = id;
	}

	public static fromProtoBufKey(
		key: string,
		options: { strict: boolean } = { strict: false },
	): ContractId {
		const out: proto.Key = proto.Key.decode(Buffer.from(key, 'hex'));
		let id =
			out?.contractID?.contractNum ||
			out?.delegatableContractId?.contractNum;
		if (options.strict && !id) {
			throw new InvalidKeyForContractIdDomainError(out);
		} else if (!id) {
			id = Long.ZERO;
		}
		return new ContractId('0.0.' + id.toString());
	}

	public toString(): string {
		throw new Error('Method not implemented.');
	}
}
