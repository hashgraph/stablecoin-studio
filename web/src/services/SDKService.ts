import { HederaNetwork, HederaNetworkEnviroment, NetworkMode, SDK } from 'hedera-stable-coin-sdk';
import type { AppMetadata, InitializationData , AcknowledgeMessage} from 'hedera-stable-coin-sdk';

export enum HashConnectConnectionState {
	Connected = 'Connected',
	Disconnected = 'Disconnected',
	Paired = 'Paired',
	Connecting = 'Connecting',
}

const appMetadata: AppMetadata = {
	name: 'dApp Example',
	description: 'An example hedera dApp',
	icon: 'https://absolute.url/to/icon.png',
	url: '',
};

interface CashInRequest {
	proxyContractId: string;
	privateKey: string;
	accountId: string;
	tokenId: string;
	targetId: string;
	amount: number;
}
interface EventsSetter {
	onInit: () => void;
	onWalletExtensionFound: () => void;
	onWalletPaired: () => void;
	onWalletAcknowledgeMessageEvent: (msg: AcknowledgeMessage) => void;
}

export class SDKService {
	private static instance: SDK | undefined;

	constructor() {}
	public static async getInstance(events?: EventsSetter) {
		if (!SDKService.instance) {
			SDKService.instance = new SDK({
				network: new HederaNetwork(HederaNetworkEnviroment.TEST), // TODO: dynamic data
				mode: NetworkMode.HASHPACK,
				options: {
					appMetadata,
				},
			});

			const { onInit, onWalletExtensionFound, onWalletPaired, onWalletAcknowledgeMessageEvent } =
				events || {
					onInit: () => {},
					onWalletAcknowledgeMessageEvent: () => {},
					onWalletExtensionFound: () => {},
					onWalletPaired: () => {},
				};

			await SDKService.instance.init({ onInit });
			SDKService.instance.onWalletExtensionFound(onWalletExtensionFound);
			SDKService.instance.onWalletPaired(onWalletPaired);
			SDKService.instance.onWalletAcknowledgeMessageEvent(onWalletAcknowledgeMessageEvent);
		}

		return SDKService.instance;
	}

	public static isInit() {
		// @ts-ignore
		return !!this.instance?.networkAdapter?._provider;
	}

	public static connectWallet() {
		SDKService.getInstance().then((instance) => instance?.connectWallet());
	}

	public static async getAvailabilityExtension(): Promise<boolean> {
		return (await SDKService.getInstance())?.getAvailabilityExtension();
	}

	public static async getStatus(): Promise<HashConnectConnectionState | undefined> {
		return (await SDKService.getInstance())?.gethashConnectConectionStatus();
	}

	public static async getWalletData(): Promise<InitializationData> {
		return (await SDKService.getInstance()).getInitData();
	}

	public static async disconnectWallet(): Promise<void> {
		return (await SDKService.getInstance()).disconectHaspack();
	}

	public static async cashIn({
		proxyContractId,
		privateKey,
		accountId,
		tokenId,
		targetId,
		amount,
	}: CashInRequest) {
		return await SDKService.getInstance().then((instance) =>
			instance.cashIn({ proxyContractId, privateKey, accountId, tokenId, targetId, amount }),
		);
	}
}

export default SDKService;
