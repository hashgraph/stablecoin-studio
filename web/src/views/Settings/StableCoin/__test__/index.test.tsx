import StableCoinSettings from '../';
import { render } from '../../../../test/index';
import translations from '../../../../translations/en/settings.json';
import userEvent from '@testing-library/user-event';
import configureMockStore from 'redux-mock-store';
import { act } from '@testing-library/react';
import SDKService from '../../../../services/SDKService';
import ContractId from '@hashgraph/stablecoin-npm-sdk/build/esm/src/domain/context/contract/ContractId';
import { Network } from '@hashgraph/stablecoin-npm-sdk';

const mockStore = configureMockStore();

describe(`<${StableCoinSettings.name} />`, () => {
	beforeEach(() => {});

	test('should render correctly', () => {
		const component = render(<StableCoinSettings />);

		expect(component.asFragment()).toMatchSnapshot();
	});

	test('should has title', () => {
		const component = render(<StableCoinSettings />);
		const header = component.getByTestId('base-container-heading');

		expect(header).toHaveTextContent(translations.title);
	});

	test('should have update owner and update implementation stablecoin buttons', async () => {
		const store = mockStore({
			wallet: {
				isProxyOwner: true,
				selectedStableCoin: {
					tokenId: '0.0.1',
					proxyAdminAddress: '0.0.2',
				},
				isAcceptOwner: false,
				isPendingOwner: false,
			},
		});

		jest.mock('react-hook-form', () => ({
			...jest.requireActual('react-hook-form'),
			Controller: () => <></>,
			useForm: () => ({
				getValues: () => ({
					updateImplementation: '0.0.12345',
					updateOwner: '0.0.12345',
				}),
			}),
		}));

		jest.spyOn(Network, 'getFactoryAddress').mockReturnValue('0.0.12345');

		jest
			.spyOn(SDKService, 'getHederaTokenManagerList')
			.mockImplementation(() => Promise.resolve([new ContractId('0.0.3')]));

		const component = render(<StableCoinSettings />, store);
		const stableCoin = await component.findByTestId('address-label');
		expect(stableCoin).toBeInTheDocument();
		expect(stableCoin).toBeEnabled();

		const selector = component.getByRole('combobox');
		await act(async () => userEvent.click(selector));
		const option = component.getAllByText('0.0.3')[1];
		userEvent.click(option);

		const addressButton = await component.findByTestId('update-implementation-address-button');
		expect(addressButton).toBeInTheDocument();
		expect(addressButton).toBeEnabled();
		await userEvent.click(addressButton);

		const newOwner = component.getByTestId('updateOwner');
		await userEvent.type(newOwner, '0.0.02468');

		const ownerButton = await component.findByTestId('update-owner-button');
		expect(ownerButton).toBeInTheDocument();
		expect(ownerButton).toBeEnabled();
		await userEvent.click(ownerButton);
	});

	test('should have accept stablecoin owner button', async () => {
		const store = mockStore({
			wallet: {
				isProxyOwner: false,
				selectedStableCoin: {
					tokenId: '0.0.1',
					proxyAdminAddress: '0.0.2',
				},
				isAcceptOwner: true,
				isPendingOwner: false,
			},
		});

		jest.mock('react-hook-form', () => ({
			...jest.requireActual('react-hook-form'),
			Controller: () => <></>,
			useForm: () => ({
				getValues: () => ({
					updateImplementation: '0.0.12345',
					updateOwner: '0.0.12345',
				}),
			}),
		}));

		jest
			.spyOn(SDKService, 'getHederaTokenManagerList')
			.mockImplementation(() => Promise.resolve([new ContractId('0.0.3')]));

		const component = render(<StableCoinSettings />, store);

		const acceptOwnerButton = await component.findByTestId('accept-owner-button');
		expect(acceptOwnerButton).toBeInTheDocument();
		expect(acceptOwnerButton).toBeEnabled();
		await userEvent.click(acceptOwnerButton);
	});

	test('should have pending stablecoin proxy owner buttons', async () => {
		const store = mockStore({
			wallet: {
				isProxyOwner: true,
				selectedStableCoin: {
					tokenId: '0.0.1',
					proxyAdminAddress: '0.0.2',
				},
				isAcceptOwner: false,
				isPendingOwner: true,
			},
		});

		jest.mock('react-hook-form', () => ({
			...jest.requireActual('react-hook-form'),
			Controller: () => <></>,
			useForm: () => ({
				getValues: () => ({
					updateImplementation: '0.0.12345',
					updateOwner: '0.0.12345',
				}),
			}),
		}));

		jest.spyOn(Network, 'getFactoryAddress').mockReturnValue('0.0.12345');

		jest
			.spyOn(SDKService, 'getHederaTokenManagerList')
			.mockImplementation(() => Promise.resolve([new ContractId('0.0.3')]));

		const component = render(<StableCoinSettings />, store);
		const stableCoin = await component.findByTestId('address-label');
		expect(stableCoin).toBeInTheDocument();
		expect(stableCoin).toBeEnabled();

		const selector = component.getByRole('combobox');
		await act(async () => userEvent.click(selector));
		const option = component.getAllByText('0.0.3')[1];
		userEvent.click(option);

		const addressButton = await component.findByTestId('update-implementation-address-button');
		expect(addressButton).toBeInTheDocument();
		expect(addressButton).toBeEnabled();
		await userEvent.click(addressButton);

		const newOwner = component.getByTestId('updateOwner');
		await userEvent.type(newOwner, '0.0.02468');

		const ownerButton = await component.findByTestId('update-owner-button');
		expect(ownerButton).toBeInTheDocument();
		expect(ownerButton).toBeEnabled();
		await userEvent.click(ownerButton);

		const cancelOwnerButton = await component.findByTestId('cancel-owner-button');
		expect(cancelOwnerButton).toBeInTheDocument();
		expect(cancelOwnerButton).toBeEnabled();
		await userEvent.click(cancelOwnerButton);
	});
});
