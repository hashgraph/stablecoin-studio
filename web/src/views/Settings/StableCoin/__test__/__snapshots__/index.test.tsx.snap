// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StableCoinSettings /> should render correctly 1`] = `
<DocumentFragment>
  <div
    class="chakra-stack css-19t9j1"
  >
    <h2
      class="chakra-heading css-bqsusc"
      data-testid="base-container-heading"
    >
      Settings
    </h2>
    <div
      class="css-1b6lda7"
      data-testid="base-container-children"
    >
      <div
        class="css-10opskl"
        data-testid="no-proof-of-reserve-container"
      >
        <div
          class="css-mnqf94"
        >
          <svg
            viewBox="0 0 2500 2500"
            width="48px"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              data-name="Layer 2"
            >
              <g
                data-name="Layer 1"
              >
                <path
                  d="M1250 0C559.64 0 0 559.64 0 1250s559.64 1250 1250 1250 1250-559.64 1250-1250S1940.36 0 1250 0"
                />
                <path
                  d="M1758.12 1790.62h-158.74v-337.49H900.62v337.49H741.87V696.25h158.75v329.37h698.76V696.25h158.75zm-850-463.75h698.75V1152.5H908.12z"
                  fill="#fff"
                />
              </g>
            </g>
          </svg>
        </div>
        <p
          class="chakra-text css-e8j0ix"
          data-testid="no-proof-of-reserve-title"
        >
          Loading data
        </p>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
